designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+* Plugin Phase 1.5 was desgined and implemented to achieve chaining of plugins. The main goal of Phase 2 plugins is Kubebuilder discovering and using external plugins, also referred as out-of-tree plugins. So Phase 2 achieves both chaining of plugins and discovery of external plugins/source code which is not intented to be shipped with the binaries. For e.g., Kubebuilder being able to discover the helm plugin and use that without having to build the plugin with the binary. These external plugins can be implemented in any language.
+
+### Related issues and PRs
+
+* [Feature Request: Plugins Phase 2](https://github.com/kubernetes-sigs/kubebuilder/issues/1378)
+* [Extensible CLI and Scaffolding Plugins - Phase 1.5](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-1-5.md)
+* [Phase 1.5 Implementation PR](https://github.com/kubernetes-sigs/kubebuilder/pull/2060)
+* [Plugin Resolution Enhancement Proposal](https://github.com/kubernetes-sigs/kubebuilder/pull/1942)
+
+### Use Cases
+
+* As a CLI user, I would like to be able to provide external source code of my plugins and its boilerplates for the CLI to be able to use them by extending the subcommands provided by Kubebuilder, which will not have the plugin binaries be shipped by default.
+
+* As a Kubebuilder/Operator SDK maintainer, I would like to be able to export an internal plugin and make that an external source, so that the plugin is no longer officially maintained by the project but still available to be used.
+  * For e.g., once phase 2 plugin implementation is completed, converting some internal plugins in OpenShift as external ones will remove the necessity to ship those plugins along with the OpenShift binary.
+
+### Goals
+
+* Kubebuilder is able to discover plugin binaries and run plugins using the CLI.
+
+* Kubebuilder can use the external plugins as its own internal ones to do scaffolding and be able to show plugin specific information such as via the --help flag.
+
+* Operator SDK is able to extend Kubebuilder and provide the same functionality for plugins chaining and enabling out-of-tree plugin discovery.
+
+### Non-Goals
+
+* Addition of new arbitrary subcommands other than the subcommands that we already support i.e init, create api, and create webhook.
+
+* Discovering plugin binaries that are not locally present on the machine (i.e. exists in a remote repository).
+
+* Discovering flags that each individual plugin would support
+  * Flags would have to be provided by the user as a part of project configuration.
+
+* Providing other options (other than stdin stdout) for inter-process communication between Kubebuilder and the plugins.
+
+### Examples
+
+* `kubebuilder create api --plugins=myexternalplugin/v1`
+  * should scaffold files using the source code of the external plugin as defined in its implementation of the create api method.
+
+* `kubebuilder create api --plugins=myexternalplugin/v1,myexternalplugin/v2`
```

Comment:
This example isn't great because you should never call the same plugin twice
```suggestion
* `kubebuilder create api --plugins=myexternalplugin/v1,myotherexternalplugin/v2`
```